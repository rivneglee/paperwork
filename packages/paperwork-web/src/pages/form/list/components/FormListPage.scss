.pwapp-form-list {
  .pw-list-item:hover {
    background: none;
  }
}
