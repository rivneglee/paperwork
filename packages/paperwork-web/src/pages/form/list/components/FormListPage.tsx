import React, { FunctionComponent } from 'react';
import {
  PaginationTemplate,
  QuickAdd,
  Search,
  Icons,
  Card,
  List,
  PageState,
} from '@paperwork/ui-widgets';
import AppBar from '../../../../components/AppBar';
import Spinner from '../../../../components/PageTransitionSpinner/Spinner';
import StickySideBar from '../../../../components/StickySideBar/StickySideBar';
import FormInfoCard, { FormItem } from './FormInfoCard';

import './FormListPage.scss';
const noResultFoundImg = require('../../../../assets/no-results-found.svg');

interface Props {
  entries: FormItem[];
  page: number;
  total: number;
  onLoadNextPage: (filterOptions: FilterOptions, page: number) => void;
  onApplyFilter: (filterOptions: FilterOptions) => void;
  onFilterChange: (option: FilterOption) => void;
  filterOptions: FilterOptions;
  isProcessing?: boolean;
  onEdit: (id: string) => void;
  onCreateNew: () => void;
}

export interface FilterOption {
  key: string;
  value: number | string;
}

export interface FilterOptions {
  keyword?: string;
}

const handleFilterChange = (key: string, handler: any) => (e: any) => handler({ key, value: e.target.value });

const FormListPage: FunctionComponent<Props> = ({
  entries,
  onLoadNextPage,
  onApplyFilter,
  onEdit,
  onCreateNew,
  onFilterChange,
  filterOptions,
  page,
  total,
  isProcessing,
}) => (
  <PaginationTemplate
    className="pwapp-form-list"
    spinner={<Spinner/>}
    isProcessing={isProcessing}
    header={<AppBar activeMenuId="forms"/>}
    onLoadMore={page => onLoadNextPage(filterOptions, page)}
    page={page}
    total={total}
  >
    <Card header={<Card.Header primary="My forms"/>}>
      <StickySideBar>
        <Search
          placeholder="Search"
          value={filterOptions.keyword}
          onChange={handleFilterChange('keyword', onFilterChange)}
          onApply={() => onApplyFilter(filterOptions)}
        />
        <QuickAdd color="secondary">
          <QuickAdd.Item onClick={onCreateNew} icon={<Icons.Form/>} tooltip="Create an empty form"/>
        </QuickAdd>
      </StickySideBar>
      {
        entries.length === 0 && (
          <PageState
            image={noResultFoundImg}
            title="No results found"
            description="Please try to improve your filter or create a new template"
          />
        )
      }
      {
        entries.length > 0 && (
          <List>
            {
              entries.map(entry => (
                <List.Item key={entry.id} onClick={() => onEdit(entry.id)}>
                  <FormInfoCard {...entry} />
                </List.Item>
              ))
            }
          </List>
        )
      }
    </Card>
  </PaginationTemplate>
);

export default FormListPage;
