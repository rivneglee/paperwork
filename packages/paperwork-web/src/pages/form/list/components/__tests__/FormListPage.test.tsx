import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import FormListPage from '../FormListPage';

describe('FormListPage', () => {
  const filterOptions = {};
  const onApplyFilter = jest.fn();
  const onFilterChange = jest.fn();
  const onLoadNextPage = jest.fn();
  const onCreateNew = jest.fn();
  const onEdit = jest.fn();
  it('should render when result is empty', () => {
    const wrapper: ShallowWrapper = shallow(
      <FormListPage
        onCreateNew={onCreateNew}
        onEdit={onEdit}
        onLoadNextPage={onLoadNextPage}
        onApplyFilter={onApplyFilter}
        onFilterChange={onFilterChange}
        entries={[]} filterOptions={filterOptions}
        page={0}
        total={10}
        isProcessing={false}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render when result is not empty', () => {
    const wrapper: ShallowWrapper = shallow(
      <FormListPage
        onLoadNextPage={onLoadNextPage}
        onApplyFilter={onApplyFilter}
        onFilterChange={onFilterChange}
        onCreateNew={onCreateNew}
        onEdit={onEdit}
        page={0}
        total={10}
        isProcessing={false}
        entries={[
          {
            closeDate: '30/07/2020',
            id: '1',
            isPublic: true,
            maxCommits: 250,
            name: 'Employee engagement survey',
            receivedCommits: 10,
            status: 'OPEN',
            statusBadgeColor: 'secondary',
            targetCommits: 100,
            createdAt: '30/06/2020',
            theme: 'purple',
            maxGap: 0,
            targetGap: 0,
            newCommitUrl: '/foo',
          },
        ]} filterOptions={filterOptions}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
