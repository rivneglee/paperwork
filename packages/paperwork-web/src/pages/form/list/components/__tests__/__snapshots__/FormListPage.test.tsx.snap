// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormListPage should render when result is empty 1`] = `
<Ai
  className="pwapp-form-list"
  header={
    <Connect(AppBar)
      activeMenuId="forms"
    />
  }
  isProcessing={false}
  onLoadMore={[Function]}
  page={0}
  spinner={<Unknown />}
  total={10}
>
  <i
    header={
      <Unknown
        primary="My forms"
      />
    }
  >
    <StickySideBar>
      <bi
        onApply={[Function]}
        onChange={[Function]}
        placeholder="Search"
      />
      <t
        color="secondary"
      >
        <Component
          icon={<Form />}
          onClick={[Function]}
          tooltip="Create an empty form"
        />
      </t>
    </StickySideBar>
    <ji
      description="Please try to improve your filter or create a new template"
      image="no-results-found.svg"
      title="No results found"
    />
  </i>
</Ai>
`;

exports[`FormListPage should render when result is not empty 1`] = `
<Ai
  className="pwapp-form-list"
  header={
    <Connect(AppBar)
      activeMenuId="forms"
    />
  }
  isProcessing={false}
  onLoadMore={[Function]}
  page={0}
  spinner={<Unknown />}
  total={10}
>
  <i
    header={
      <Unknown
        primary="My forms"
      />
    }
  >
    <StickySideBar>
      <bi
        onApply={[Function]}
        onChange={[Function]}
        placeholder="Search"
      />
      <t
        color="secondary"
      >
        <Component
          icon={<Form />}
          onClick={[Function]}
          tooltip="Create an empty form"
        />
      </t>
    </StickySideBar>
    <u>
      <Component
        key="1"
      >
        <FormInfoCard
          closeDate="30/07/2020"
          createdAt="30/06/2020"
          id="1"
          isPublic={true}
          maxCommits={250}
          maxGap={0}
          name="Employee engagement survey"
          newCommitUrl="/foo"
          onEdit={[Function]}
          receivedCommits={10}
          status="OPEN"
          statusBadgeColor="secondary"
          targetCommits={100}
          targetGap={0}
          theme="purple"
        />
      </Component>
    </u>
  </i>
</Ai>
`;
